rfcs/038-unify-shared-dependencies.md

```diff
@@ -0,0 +1,112 @@
+
+# Unify shared dependencies
+
+## Summary
+
+Major motivation behind migrating to monorepo was to solve `Dependency Management` issue. However, it still remains unsolved.
+
+This RFC serves as a platform to figure out the best approach for unifying shared dependencies and contains preliminary proposal for how that could be solved.
+
+## Motivation
+
+Following original RFC:
+
+```
+Dependency Management: Python packages are pretty much the same for each proxy, which makes it hard to sync across all the above repositories. As a result most of the time, each proxy is running its own version of the dependency.
+A few examples:
+
+| Package             | amundsenmetadata   | amundsenfrontend  |
+| --------------      |:------------------:| -----------------:|
+| amundsen-common     | >=0.8.1            | ==0.6.0           |
+| flake8              | ==3.5.0            | ==3.8.4           |
+| flake8-tidy-imports | ==1.1.0            | ==4.2.1           |
+
+and many more... This change will put all the packages in just one place, making practically to keep dependencies similar accross verisons, simply because it will be possible to open a single PR that updates multiple packages. Practically speaking this will improve the status quo slightly.
+```
+
+## Guide-level Explanation (aka Product Details)
+
+n/a
+
+## UI/UX-level Explanation
+
+n/a
+
+## Reference-level Explanation (aka Technical Details)
+
+Following steps are proposed to achieve goal of this RFC. 
+
+Multiple times you might find the term `subdirectory`, which refers to following folders:
+- common
+- databuilder
+- frontend
+- metadata
+- search
+
+Steps:
+
+1. Split `requirements.txt` for each subdirectory into `requirements.txt` and `requirements-dev.txt` (there is no need to install flake8 when doing for example `pip install amundsen-metadata` which is currently the case)
+2. Identify libraries that are shared across all (or majority - see `Drawbacks`) of subprojects
+3. Create `requiremnts-common.txt` file in root folder of amundsen repo containing pinned dependencies for shared libraries
+4. Remove libraries listed in `requirements-common.txt` from `requirements.txt` of each subdirectory
+5. Create `requirements-dev.txt` file in root folder of amundsen repo containing pinned dependencies for shared dev libraries (mypy, flake8, etc.)
+6. Remove libraries listed in `requirements-dev.txt` from `requirements.txt` of each subdirectory
+7. Create symbolic link to `requirements-common.txt` and `requirements-dev.txt` in each subdirectory
+8. Shift context of building docker images to root folder (otherwise we cannot share `requirements-common.txt` and `requirements-dev.txt` symlinks across subdirectories)
+9. Extend every `setup.py` to contain following entries (`extras_require` might contain more keys)"
+```python
+install_require=requirements + requirements_common
```

Comment:
For Databuilder, I think this is less-than-ideal, since we expect it to actually be installed as a library, and thus having precisely pinned deps in setup.py might cause issues. For the other "libraries", I don't think it's as a big a deal, since they are generally used standalone. wdyt?

Reply:
from my brief experiments I've actually identified databuilder as more `problematic` - mostly cause of the multitude of optional extractors (and hence > a lot of sections in `extras_require`). We could be good by partially removing it from the approach - so keep the requirements.txt as is, do not include `requirements-common.txt` but use shared `requirements-dev.txt`

Reply:
sgtm to exclude it.